File: easy/array/duplicates_2/duplicates_2.go
package main

import "fmt"

func containsNearbyDuplicate(nums []int, k int) bool {
	numbersToIndex := make(map[int]int, len(nums))
	for i, n := range nums {
		if prevIndex, ok := numbersToIndex[n]; ok && i-prevIndex <= k {
			return true
		}
		numbersToIndex[n] = i
	}
	return false
}

func main() {
	fmt.Println("Solution 1 : ", containsNearbyDuplicate([]int{1, 2, 3, 1}, 3))
	fmt.Println("Solution 2 : ", containsNearbyDuplicate([]int{1, 0, 1, 1}, 1))
	fmt.Println("Solution 3 : ", containsNearbyDuplicate([]int{1, 2, 3, 1, 2, 3}, 2))
}

File: easy/array/duplicates_2/duplicates_2.py
from typing import List


class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        number_to_index = {}
        for i, n in enumerate(nums):
            if n in number_to_index and i-number_to_index[n] <= k:
                return True
            number_to_index[n] = i
        return False


s = Solution()
print("Solution 1 : ", s.containsNearbyDuplicate([1, 2, 3, 1], 3))
print("Solution 2 : ", s.containsNearbyDuplicate([1, 0, 1, 1], 1))
print("Solution 3 : ", s.containsNearbyDuplicate([1, 2, 3, 1, 2, 3], 2))
